[JSC][32bit] Enable wasm signaling memory

mikhailramalho opened the pull request:
#### ad17368e4bdbc6099f02e6c3871b05926893bcea
<pre>
[JSC][32bit] Enable wasm signaling memory
<a href="https://bugs.webkit.org/show_bug.cgi?id=244843">https://bugs.webkit.org/show_bug.cgi?id=244843</a>

Reviewed by Yusuke Suzuki.

Enable wasm signaling memory on 32bit architectures. The only thing we
need to do is to limit the &quot;fast mapped memory&quot; size to 2GB, that way
we can fit that plus the redzone pages inside a 32bit pointer. This is
OK because our memory limit was already 2GB anyway
(MAX_ARRAY_BUFFER_SIZE).

* JSTests/wasm.yaml: unskip shared memory tests on arm.
* JSTests/wasm/stress/shared-memory-errors.js: ditto.
* JSTests/wasm/stress/shared-wasm-memory-buffer.js: ditto.
* Source/JavaScriptCore/wasm/WasmMemory.cpp: use a fast memory mapped
* size of 2GB for 32bit architectures.
* (JSC::Wasm::Memory::fastMappedBytes):
* Source/WTF/wtf/PlatformEnable.h: enable wasm memory signaling on
* 32bit.
* Source/JavaScriptCore/runtime/BufferMemoryHandle.cpp:
(JSC::BufferMemoryHandle::fastMappedRedzoneBytes):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):

Canonical link: <a href="https://commits.webkit.org/260483@main">https://commits.webkit.org/260483@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/109c9195969768357d0dafadf2597543149e4811

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108439 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17535 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41289 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117544 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18986 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8811 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100663 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114207 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97448 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42185 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97595 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10352 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30437 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/96985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8471 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11104 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/96985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16499 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50033 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106016 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7250 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12690 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26235 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me. Please ensure that fast memory is not enabled. Wasm fast memory is not available on 32bit, but signaling memory (used for shared memory with bound checking) is available.

Constellation commented on Source/JavaScriptCore/runtime/Options.cpp:
```diff
@@ -751,6 +751,8 @@ void Options::notifyOptionsChanged()
         Options::useWebAssemblyFastMemory() = false;
 
+    Options::useWebAssemblyFastMemory() = canUseWebAssemblyFastMemory();
+
```
This is wrong. It will break `= false` test case.
Do,

```
#CPU(ADDRESS32)
Options::useWebAssemblyFastMemory() = false
#end
```

Constellation commented on Source/JavaScriptCore/runtime/BufferMemoryHandle.cpp:
```diff
@@ -51,13 +51,18 @@ namespace JSC {
-    return static_cast<size_t>(PageCount::pageSize) * Options::webAssemblyFastMemoryRedzonePages();
+    if (canUseWebAssemblyFastMemory())
+        return static_cast<size_t>(PageCount::pageSize) * Options::webAssemblyFastMemoryRedzonePages();
+    return 0;
```
This is not correct. If this code is used, since this is *fast* memory code, it should return the correct size. So, do not change this.

Constellation reviewed (approved):
r=me

xanlpz committed:
[JSC][32bit] Enable wasm signaling memory
https://bugs.webkit.org/show_bug.cgi?id=244843

Reviewed by Yusuke Suzuki.

Enable wasm signaling memory on 32bit architectures. The only thing we
need to do is to limit the "fast mapped memory" size to 2GB, that way
we can fit that plus the redzone pages inside a 32bit pointer. This is
OK because our memory limit was already 2GB anyway
(MAX_ARRAY_BUFFER_SIZE).

* JSTests/wasm.yaml: unskip shared memory tests on arm.
* JSTests/wasm/stress/shared-memory-errors.js: ditto.
* JSTests/wasm/stress/shared-wasm-memory-buffer.js: ditto.
* Source/JavaScriptCore/wasm/WasmMemory.cpp: use a fast memory mapped
* size of 2GB for 32bit architectures.
* (JSC::Wasm::Memory::fastMappedBytes):
* Source/WTF/wtf/PlatformEnable.h: enable wasm memory signaling on
* 32bit.
* Source/JavaScriptCore/runtime/BufferMemoryHandle.cpp:
(JSC::BufferMemoryHandle::fastMappedRedzoneBytes):
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):

Canonical link: https://commits.webkit.org/260483@main
